docs/RFCS/20190906_flow_control.md

```diff
@@ -0,0 +1,131 @@
+# Proposal: Add flow control for data replication to downstream
+
+- Author(s):    [yangfei](https://github.com/amyangfei)
+- Last updated: 2019-09-18
+
+## Abstract
+
+Too fast import data or replication in DM may lead to downstream TiDB too busy to serve as normal. This proposal aims to find a way to detect the congestion of import or replication data flow and introduces a framework of dynamic concurrency control, which will help to control the import and replication data flow.
+
+Table of contents:
+
+- [Background](#Background)
+- [Implementation](#Implementation)
+    - [Congestion definition](#Congestion-definition)
+        - [How to measure the key indices](#How-to-measure-the-key-indices)
+        - [How to detect and measure congestion](#How-to-detect-and-measure-congestion)
+    - [Congestion control framework](#Congestion-control-framework)
+        - [The congestion detect component](#The-congestion-detect-component)
+        - [The data flow control component](#The-data-flow-control-component)
+        - [The concurrency control component](#The-concurrency-control-component)
+
+## Background
+
+Firstly we will discuss the key factor that affects the import or replication speed between DM data processing unit and downstream TiDB.
+
+- In the full data import procedure, data import speed is based on the SQL batch size and SQL executor concurrency. The SQL batch size is mainly determined by the single statement size of dump data, which can be controlled with the `mydumper` configuration. The SQL executor concurrency equals to the worker count of load unit, which is configurated by loader `pool-size` in task config.
+- In the incremental data replication procedure, replication speed relates to SQL job batch size and SQL executor concurrency. The binlog replication unit gets data from relay log consumer (assuming the relay log consume speed is fast enough), and distributes SQL jobs to multiple executors, the executor count is configurated by `worker-count` in task config. When the cached SQL jobs' count exceed SQL batch size, all SQLs in these jobs will be executed to downstream in a transaction.
+
+In the full data import scenario, it is easy to meet downstream congestion if we have many DM-workers and much high executor concurrency. While in incremental data replication scenario the congestion does not happen often, but sometimes if user pauses replication task for a while and a lot of relay log is accumulated or the replication nodes too many with high concurrency, it is also possible to meet downstream congestion. Besides too many upstream shard instances can also cause downstream congestion. When we encounter these abnormal scenario, we often pause part of the tasks or decrease the SQL executor concurrency. The key point of solving the congestion problem is to reduce the data flow speed.
```

Comment:
> it is easy to meet downstream congestion if we have many DM-workers and much high executor concurrency

Is `easy` (not difficult) suitable here? 

> the replication nodes too many with high concurrency

how about `too many replication nodes with high concurrency`

> too many upstream shard instances can also cause downstream congestion.

no-shard instances (like too many tasks from multiple instances) can also lead to this.

---

maybe you can use an _unordered list_ to list these cases?